pkg/localstorev2/internal/reserve/sampler.go

```diff
@@ -0,0 +1,220 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package reserve
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"encoding/binary"
+	"fmt"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/bmtpool"
+	"github.com/ethersphere/bee/pkg/localstorev2/internal"
+	storage "github.com/ethersphere/bee/pkg/storagev2"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"go.uber.org/atomic"
+	"golang.org/x/sync/errgroup"
+)
+
+const sampleSize = 8
+
+type sampleStat struct {
+	TotalIterated      atomic.Int64
+	NotFound           atomic.Int64
+	NewIgnored         atomic.Int64
+	IterationDuration  atomic.Int64
+	GetDuration        atomic.Int64
+	HmacrDuration      atomic.Int64
+	ValidStampDuration atomic.Int64
+}
+
+type sampleEntry struct {
+	transformedAddress swarm.Address
+	chunk              swarm.Chunk
+}
+
+func (s sampleStat) String() string {
+
+	seconds := int64(time.Second)
+
+	return fmt.Sprintf(
+		"Chunks: %d NotFound: %d New Ignored: %d Iteration Duration: %d secs GetDuration: %d secs"+
+			" HmacrDuration: %d secs ValidStampDuration: %d secs",
+		s.TotalIterated.Load(),
+		s.NotFound.Load(),
+		s.NewIgnored.Load(),
+		s.IterationDuration.Load()/seconds,
+		s.GetDuration.Load()/seconds,
+		s.HmacrDuration.Load()/seconds,
+		s.ValidStampDuration.Load()/seconds,
+	)
+}
+
+// TODO
+func (r *Reserve) ReserveSample(
+	ctx context.Context,
+	store internal.Storage,
+	anchor []byte,
+	storageRadius uint8,
+	consensusTime uint64,
+) (Sample, error) {
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	addrChan := make(chan swarm.Address)
+	var stat sampleStat
+
+	indexStore := store.IndexStore()
+	chunkStore := store.ChunkStore()
+
+	t := time.Now()
+
+	// Phase 1: Iterate chunk addresses
+	g.Go(func() error {
+		defer close(addrChan)
+		iterationStart := time.Now()
+
+		err := indexStore.Iterate(storage.Query{
```

Comment:
So this is the only thing we need really for the sampler from the internal/reserve package. I think it would be better to have the sampler code in the main localstorev2 package by creating a function for this in internal/reserve.

Mainly because I dont think of this as being internal code. Also, in the next phase, we will have inclusion proofs which are in the `bmt` package. So there might be some interaction there. Also, there were ideas about using a different hmacr.